docs/contributing/effects-tutorial.md

```diff
@@ -0,0 +1,634 @@
+doTheProgram conf seenValues = do
+  -- Get the value from the file
+  -- Throws an error if the file is missing
+  contents <- readFile $ takeConfig conf
+
+  -- Use the magic reader function, with our wrapper
+  -- We're matching on a (Maybe ValidValue)
+  case extractValue' contents of
+    -- Terminate the recursion
+    Nothing -> pure []
+    -- We have a valid value
+    Just value -> do
+      --Have we seen this before
+      let isNewValue = value `Set.member` seenValues
+      -- Don't let this value repeat itself
+      -- Idempotent, so we don't have to check
+      let seenValues' = insert value seenValues
+      if isNewValue
+        then do -- We have a new value, we'll add it to our list builder
+          sleepForOneSecond
+          -- Recurse into the next read operation
+          values <- doTheProgram conf seenValues'
+          -- Once recursion has terminated, return the list builder
+          pure (value : values)
+        else do -- We've already seen this value, just reread
+          sleepForOneSecond
+          doTheProgram conf seenValues'
+```
+
+One notable issue is that fact that we have to carry around all of our data
+explicitly.  We could build a record type with some helper functions to manage
+all of the data, but we'd still have to carry it around everywhere.
+
+This is not always a problem, and is easily manageable in a program this small,
+but we'll see how that can be trivially solved with monad transformers and
+`fused-effects`.
+
+Let's try it out with Monad Transformers:
+
+```haskell
+main :: IO ()
+main = do
+  conf <- readConfig
+  -- The transformers have really inconvenient argument orderings, but we'll ignore that.
+  values <- execWriterT $ (`evalStateT` mempty) $ (`runReaderT` conf) doTheProgram
+  traverse_ (putStrLn . takeValue) values
+
+doTheProgram :: 
+  ( MonadReader AppConfig m
+  , MonadWriter [ValidValue] m
+  , MonadState SeenValues m
+  , MonadIO m
+  ) => 
+  m ()
+doTheProgram = do
+  -- Get the config from the reader
+  filepath <- asks takeConfig
+  -- Read the actual file
+  contents <- liftIO $ readFile filepath
+  case extractValue contents of
+    -- Value is invalid, terminate
+    Nothing -> pure ()
+    Just value -> do
+      -- Check if the value existed previously
+      isNewValue <- gets (Set.member value)
+      -- We've seen this value before
+      modify (Set.insert value)
+      if isNewValue
+        then do
+          -- This is a new value
+          tell [value]
+          liftIO sleepForOneSecond
+          doTheProgram
+        else doTheProgram
+```
+
+`mtl`-style monad transformers are quite fast, compared to other, more generic
+solutions, and are on par with the non-`mtl` version.  The _reason_ for this is
+that the `m` in the return value of `doTheProgram` is generic at the definition
+site, but when we call it with an explicit list of monad runners in our `main`
+function, that generic `m` is monomorphized to
```

Comment:
I worry that this term "monomorphized" might be a surprise to some readers.  Also that "remember, generics are slow" feels like it comes out of nowhere because you haven't talked much about performance yet, I think.